## Intro a Machine Learning 

Empezamos una nueva lección, y este es el momento de poner en práctica los conocimientos adquiridos sobre la Regresión Lineal Simple I. Para ello os vamos a pedir que por vuestra cuenta busqueis un dataset que os guste, por ejemplo de Kaggle o de donde queráis extraerlo. 
De forma que vamos a probar a realizar ajustes lineales con el fin de predecir alguna de sus variables. 
Este dataset será el que iremos usando a lo largo de todas las sesiones de pairprogramming de regresión lineal.

Objetivos
Los objetivos de hoy son:
* Buscar un conjunto de datos a analizar:
Es necesario que el conjunto de datos a analizar tenga variables numéricas y categóricas, primando que haya más de una variable de tipo numérico. Ya que tambien se utilizarán serán útiles a la hora de realizar predicciones en las clases de regresión lineal múltiple. Así mismo se recomienda que los datasets esten ligeramente limpios para facilidar los análisis.
* Explicar los datos y las variables disponibles en el conjunto de datos seleccionado:
Esto se hace para que comprendais mejor los datos con los que estais trabajando.
Realizar un EDA sencillo para la variable dependiente y la variable independiente.

In [1]:
# tratamiento de los datos

import pandas as pd
import numpy as np

# librerías para la visualización de los datos

import matplotlib.pyplot as plt
import seaborn as sns

# Establecer tamaño fijo de gráficas

plt.rcParams["figure.figsize"] = (10,8)

# Configuración warnings

import warnings
warnings.filterwarnings('ignore')

#Configuración para poder ver todas las columnas
pd.options.display.max_columns=None

In [2]:
df = pd.read_csv('../Datos/marketing_campaign.csv', sep='\t')
df.head(15)

ID  Year_Birth   Education Marital_Status   Income  Kidhome  Teenhome  \
0   5524        1957  Graduation         Single  58138.0        0         0   
1   2174        1954  Graduation         Single  46344.0        1         1   
2   4141        1965  Graduation       Together  71613.0        0         0   
3   6182        1984  Graduation       Together  26646.0        1         0   
4   5324        1981         PhD        Married  58293.0        1         0   
5   7446        1967      Master       Together  62513.0        0         1   
6    965        1971  Graduation       Divorced  55635.0        0         1   
7   6177        1985         PhD        Married  33454.0        1         0   
8   4855        1974         PhD       Together  30351.0        1         0   
9   5899        1950         PhD       Together   5648.0        1         1   
10  1994        1983  Graduation        Married      NaN        1         0   
11   387        1976       Basic        Married   7500.0        0         0   
12  2125        1959  Graduation       Divorced  63033.0        0         0   
13  8180        1952      Master       Divorced  59354.0        1         1   
14  2569        1987  Graduation        Married  17323.0        0         0   

   Dt_Customer  Recency  MntWines  MntFruits  MntMeatProducts  \
0   04-09-2012       58       635         88              546   
1   08-03-2014       38        11          1                6   
2   21-08-2013       26       426         49              127   
3   10-02-2014       26        11          4               20   
4   19-01-2014       94       173         43              118   
5   09-09-2013       16       520         42               98   
6   13-11-2012       34       235         65              164   
7   08-05-2013       32        76         10               56   
8   06-06-2013       19        14          0               24   
9   13-03-2014       68        28          0                6   
10  15-11-2013       11         5          5                6   
11  13-11-2012       59         6         16               11   
12  15-11-2013       82       194         61              480   
13  15-11-2013       53       233          2               53   
14  10-10-2012       38         3         14               17   

    MntFishProducts  MntSweetProducts  MntGoldProds  NumDealsPurchases  \
0               172                88            88                  3   
1                 2                 1             6                  2   
2               111                21            42                  1   
3                10                 3             5                  2   
4                46                27            15                  5   
5                 0                42            14                  2   
6                50                49            27                  4   
7                 3                 1            23                  2   
8                 3                 3             2                  1   
9                 1                 1            13                  1   
10                0                 2             1                  1   
11               11                 1            16                  1   
12              225               112            30                  1   
13                3                 5            14                  3   
14                6                 1             5                  1   

    NumWebPurchases  NumCatalogPurchases  NumStorePurchases  \
0                 8                   10                  4   
1                 1                    1                  2   
2                 8                    2                 10   
3                 2                    0                  4   
4                 5                    3                  6   
5                 6                    4                 10   
6                 7                    3                  7   
7          

` Explicación de las variables contenidas dentro de nuestro dataset `

**People:**
* ID: Customer's unique identifier
* Year_Birth: Customer's birth year
* Education: Customer's education level
* Marital_Status: Customer's marital status
* Income: Customer's yearly household income
* Kidhome: Number of children in customer's household
* Teenhome: Number of teenagers in customer's household
* Dt_Customer: Date of customer's enrollment with the company
* Recency: Number of days since customer's last purchase
* Complain: 1 if the customer complained in the last 2 years, 0 otherwise

**Products:**
* MntWines: Amount spent on wine in last 2 years
* MntFruits: Amount spent on fruits in last 2 years
* MntMeatProducts: Amount spent on meat in last 2 years
* MntFishProducts: Amount spent on fish in last 2 years
* MntSweetProducts: Amount spent on sweets in last 2 years
* MntGoldProds: Amount spent on gold in last 2 years

**Promotion:**
* NumDealsPurchases: Number of purchases made with a discount
* AcceptedCmp1: 1 if customer accepted the offer in the 1st campaign, 0 otherwise
* AcceptedCmp2: 1 if customer accepted the offer in the 2nd campaign, 0 otherwise
* AcceptedCmp3: 1 if customer accepted the offer in the 3rd campaign, 0 otherwise
* AcceptedCmp4: 1 if customer accepted the offer in the 4th campaign, 0 otherwise
* AcceptedCmp5: 1 if customer accepted the offer in the 5th campaign, 0 otherwise
* Response: 1 if customer accepted the offer in the last campaign, 0 otherwise

**Place:**
* NumWebPurchases: Number of purchases made through the company’s website
* NumCatalogPurchases: Number of purchases made using a catalogue
* NumStorePurchases: Number of purchases made directly in stores
* NumWebVisitsMonth: Number of visits to company’s website in the last month

`Eda`

In [3]:
# Vamos a ver la forma de nuestro dataframe
df.shape

(2240, 29)

In [4]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2240 entries, 0 to 2239
Data columns (total 29 columns):
 #   Column               Non-Null Count  Dtype  
---  ------               --------------  -----  
 0   ID                   2240 non-null   int64  
 1   Year_Birth           2240 non-null   int64  
 2   Education            2240 non-null   object 
 3   Marital_Status       2240 non-null   object 
 4   Income               2216 non-null   float64
 5   Kidhome              2240 non-null   int64  
 6   Teenhome             2240 non-null   int64  
 7   Dt_Customer          2240 non-null   object 
 8   Recency              2240 non-null   int64  
 9   MntWines             2240 non-null   int64  
 10  MntFruits            2240 non-null   int64  
 11  MntMeatProducts      2240 non-null   int64  
 12  MntFishProducts      2240 non-null   int64  
 13  MntSweetProducts     2240 non-null   int64  
 14  MntGoldProds         2240 non-null   int64  
 15  NumDealsPurchases    2240 non-null   i

> Observamos que sólo tenemos nulos en 'income' y que los tipos de datos parecen correctos.

In [5]:
# Vamos a calcular los principales cálculos estadísticos para las columnas numéricas
df.describe().T

count          mean           std     min       25%  \
ID                   2240.0   5592.159821   3246.662198     0.0   2828.25   
Year_Birth           2240.0   1968.805804     11.984069  1893.0   1959.00   
Income               2216.0  52247.251354  25173.076661  1730.0  35303.00   
Kidhome              2240.0      0.444196      0.538398     0.0      0.00   
Teenhome             2240.0      0.506250      0.544538     0.0      0.00   
Recency              2240.0     49.109375     28.962453     0.0     24.00   
MntWines             2240.0    303.935714    336.597393     0.0     23.75   
MntFruits            2240.0     26.302232     39.773434     0.0      1.00   
MntMeatProducts      2240.0    166.950000    225.715373     0.0     16.00   
MntFishProducts      2240.0     37.525446     54.628979     0.0      3.00   
MntSweetProducts     2240.0     27.062946     41.280498     0.0      1.00   
MntGoldProds         2240.0     44.021875     52.167439     0.0      9.00   
NumDealsPurchases    2240.0      2.325000      1.932238     0.0      1.00   
NumWebPurchases      2240.0      4.084821      2.778714     0.0      2.00   
NumCatalogPurchases  2240.0      2.662054      2.923101     0.0      0.00   
NumStorePurchases    2240.0      5.790179      3.250958     0.0      3.00   
NumWebVisitsMonth    2240.0      5.316518      2.426645     0.0      3.00   
AcceptedCmp3         2240.0      0.072768      0.259813     0.0      0.00   
AcceptedCmp4         2240.0      0.074554      0.262728     0.0      0.00   
AcceptedCmp5         2240.0      0.072768      0.259813     0.0      0.00   
AcceptedCmp1         2240.0      0.064286      0.245316     0.0      0.00   
AcceptedCmp2         2240.0      0.013393      0.114976     0.0      0.00   
Complain             2240.0      0.009375      0.096391     0.0      0.00   
Z_CostContact        2240.0      3.000000      0.000000     3.0      3.00   
Z_Revenue            2240.0     11.000000      0.000000    11.0     11.00   
Response             2240.0      0.149107      0.356274     0.0      0.00   

                         50%       75%       max  
ID                    5458.5   8427.75   11191.0  
Year_Birth            1970.0   1977.00    1996.0  
Income               51381.5  68522.00  666666.0  
Kidhome                  0.0      1.00       2.0  
Teenhome                 0.0      1.00       2.0  
Recency                 49.0     74.00      99.0  
MntWines               173.5    504.25    1493.0  
MntFruits                8.0     33.00     199.0  
MntMeatProducts         67.0    232.00    1725.0  
MntFishProducts         12.0     50.00     259.0  
MntSweetProducts         8.0     33.00     263.0  
MntGoldProds            24.0     56.00     362.0  
NumDealsPurchases        2.0      3.00      15.0  
NumWebPurchases          4.0      6.00      27.0  
NumCatalogPurchases      2.0      4.00      28.0  
NumStorePurchases        5.0      8.00      13.0  
NumWebVisitsMonth        6.0      7.00      20.0  
AcceptedCmp3             0.0      0.00       1.0  
AcceptedCmp4             0.0      0.00       1.0  
AcceptedCmp5             0.0      0.00       1.0  
AcceptedCmp1             0.0      0.00       1.0  
AcceptedCmp2             0.0      0.00       1.0  
Complain                 0.0      0.00       1.0  
Z_CostContact            3.0      3.00       3.0  
Z_Revenue               11.0     11.00      11.0  
Response                 0.0      0.00       1.0

> Vemos que la mayoría de personas en nuestro dataset nacieron en 1970, no tienen hijos pequeños ni adolescentes y consumen $173 en vino cada dos años de mediana.

In [6]:
# Vamos a calcular los principales cálculos estadísticos para las columnas categóricas
df.describe(include='O').T

count unique         top  freq
Education       2240      5  Graduation  1127
Marital_Status  2240      8     Married   864
Dt_Customer     2240    663  31-08-2012    12

> La mayoría tienen un nivel de educación universitario y están casados.
>
> La fecha en la que se registraron más clientes fue 31-08-2012

In [65]:
# Vemos los nulos de forma más clara
df.isnull().sum()

ID                      0
Year_Birth              0
Education               0
Marital_Status          0
Income                 24
Kidhome                 0
Teenhome                0
Dt_Customer             0
Recency                 0
MntWines                0
MntFruits               0
MntMeatProducts         0
MntFishProducts         0
MntSweetProducts        0
MntGoldProds            0
NumDealsPurchases       0
NumWebPurchases         0
NumCatalogPurchases     0
NumStorePurchases       0
NumWebVisitsMonth       0
AcceptedCmp3            0
AcceptedCmp4            0
AcceptedCmp5            0
AcceptedCmp1            0
AcceptedCmp2            0
Complain                0
Z_CostContact           0
Z_Revenue               0
Response                0
dtype: int64

Vamos a hacer un dataframe con las frecuencias de las variables categóricas para ver si necesitamos hacer limpieza de las categorías

In [7]:
# Separamos variables numéricas de categóricas
numericas= df.select_dtypes(include=np.number)
categoricas= df.select_dtypes(include='object')

In [8]:
categoricas.head()

Education Marital_Status Dt_Customer
0  Graduation         Single  04-09-2012
1  Graduation         Single  08-03-2014
2  Graduation       Together  21-08-2013
3  Graduation       Together  10-02-2014
4         PhD        Married  19-01-2014

In [9]:
# Creamos nuestro diccionario de frecuencias para luego hacer un dataframe
dicc = {'variable':[], 'valores_unicos':[], 'frecuencias':[]}

for col in categoricas.columns:
    dicc['variable'].append(col)
    dicc['valores_unicos'].append(list(categoricas[col].value_counts().index))
    dicc['frecuencias'].append(list(categoricas[col].value_counts().values))

In [10]:
df_uniques = pd.DataFrame(dicc)
df_uniques

variable                                     valores_unicos  \
0       Education         [Graduation, PhD, Master, 2n Cycle, Basic]   
1  Marital_Status  [Married, Together, Single, Divorced, Widow, A...   
2     Dt_Customer  [31-08-2012, 12-09-2012, 14-02-2013, 12-05-201...   

                                         frecuencias  
0                          [1127, 486, 370, 203, 54]  
1                  [864, 580, 480, 232, 77, 3, 2, 2]  
2  [12, 11, 11, 11, 10, 10, 9, 9, 9, 9, 9, 8, 8, ...

Vemos que la columna 'Dt_customer' no nos va a ser útil para nuestro análisis, ya que no aporta información relevante y tiene demasiados valores únicos

In [11]:
# Eliminamos la columna 'Dt_customer'
df.drop('Dt_Customer', axis=1, inplace=True)

Renombramos las columnas.

In [12]:
nuevas_columnas = {col: col.lower() for col in df.columns}
df.rename(columns = nuevas_columnas, inplace = True)

Crearemos una nueva columna con el total gastado por cada cliente en otras categorías, sumando las columnas de montos gastados.

In [75]:
# Miramos los nombres de las variables que nos interesan
df.columns

Index(['id', 'year_birth', 'education', 'marital_status', 'income', 'kidhome',
       'teenhome', 'dt_customer', 'recency', 'mntwines', 'mntfruits',
       'mntmeatproducts', 'mntfishproducts', 'mntsweetproducts',
       'mntgoldprods', 'numdealspurchases', 'numwebpurchases',
       'numcatalogpurchases', 'numstorepurchases', 'numwebvisitsmonth',
       'acceptedcmp3', 'acceptedcmp4', 'acceptedcmp5', 'acceptedcmp1',
       'acceptedcmp2', 'complain', 'z_costcontact', 'z_revenue', 'response'],
      dtype='object')

In [13]:
# Sumamos sus valores para obtener el total gastado en productos excepto en vino
df['mnt_total_sin_vino'] = df ['mntfruits'] + df['mntmeatproducts'] + df['mntfishproducts'] + df['mntsweetproducts'] + df['mntgoldprods']

In [14]:
df.head()

id  year_birth   education marital_status   income  kidhome  teenhome  \
0  5524        1957  Graduation         Single  58138.0        0         0   
1  2174        1954  Graduation         Single  46344.0        1         1   
2  4141        1965  Graduation       Together  71613.0        0         0   
3  6182        1984  Graduation       Together  26646.0        1         0   
4  5324        1981         PhD        Married  58293.0        1         0   

   recency  mntwines  mntfruits  mntmeatproducts  mntfishproducts  \
0       58       635         88              546              172   
1       38        11          1                6                2   
2       26       426         49              127              111   
3       26        11          4               20               10   
4       94       173         43              118               46   

   mntsweetproducts  mntgoldprods  numdealspurchases  numwebpurchases  \
0                88            88                  3                8   
1                 1             6                  2                1   
2                21            42                  1                8   
3                 3             5                  2                2   
4                27            15                  5                5   

   numcatalogpurchases  numstorepurchases  numwebvisitsmonth  acceptedcmp3  \
0                   10                  4                  7             0   
1                    1                  2                  5             0   
2                    2                 10                  4             0   
3                    0                  4                  6             0   
4                    3                  6                  5             0   

   acceptedcmp4  acceptedcmp5  acceptedcmp1  acceptedcmp2  complain  \
0             0             0             0             0         0   
1             0             0             0             0         0   
2             0             0             0             0         0   
3             0             0             0             0         0   
4             0             0             0             0         0   

   z_costcontact  z_revenue  response  mnt_total_sin_vino  
0              3         11         1                 982  
1              3         11         0                  16  
2              3         11         0                 350  
3              3         11         0                  42  
4              3         11         0                 249

Vamos a observar nuestras columnas numéricas

In [15]:
numericas.head()

ID  Year_Birth   Income  Kidhome  Teenhome  Recency  MntWines  MntFruits  \
0  5524        1957  58138.0        0         0       58       635         88   
1  2174        1954  46344.0        1         1       38        11          1   
2  4141        1965  71613.0        0         0       26       426         49   
3  6182        1984  26646.0        1         0       26        11          4   
4  5324        1981  58293.0        1         0       94       173         43   

   MntMeatProducts  MntFishProducts  MntSweetProducts  MntGoldProds  \
0              546              172                88            88   
1                6                2                 1             6   
2              127              111                21            42   
3               20               10                 3             5   
4              118               46                27            15   

   NumDealsPurchases  NumWebPurchases  NumCatalogPurchases  NumStorePurchases  \
0                  3                8                   10                  4   
1                  2                1                    1                  2   
2                  1                8                    2                 10   
3                  2                2                    0                  4   
4                  5                5                    3                  6   

   NumWebVisitsMonth  AcceptedCmp3  AcceptedCmp4  AcceptedCmp5  AcceptedCmp1  \
0                  7             0             0             0             0   
1                  5             0             0             0             0   
2                  4             0             0             0             0   
3                  6             0             0             0             0   
4                  5             0             0             0             0   

   AcceptedCmp2  Complain  Z_CostContact  Z_Revenue  Response  
0             0         0              3         11         1  
1             0         0              3         11         0  
2             0         0              3         11         0  
3             0         0              3         11         0  
4             0         0              3         11         0

Hemos detectado varias columnas que no nos van a ser de utilidad para nuestro análisis, ya que no tenemos claro su significado y no aportan mucha información relevante

In [78]:
df.z_costcontact.unique()

array([3])

In [79]:
df.z_revenue.unique()

array([11])

Puesto que nuestro análisis no va dirigido a analizar campañas de marketing, incluimos también las columnas que nos indican los clientes que compraron en distintas campañas de marketing y las eliminaremos también.

In [16]:
df.drop(['z_costcontact', 'z_revenue','acceptedcmp3', 'acceptedcmp4', 'acceptedcmp5', 'acceptedcmp1', 'acceptedcmp2', 'complain', 'response'], axis=1, inplace=True)

Gestión de nulos: Aunque son pocos, hemos decidido sustituir los valores nulos de la columna `income` con la mediana en vez de eliminarlos, y así no perdemos los datos que nos puedan aportar esos registros

In [17]:
# Vemos que su tipo de dato es float, señal de que hay nulos
df['income'].dtypes

dtype('float64')

In [18]:
df['income'].replace(np.nan, df['income'].median(), inplace = True)

In [20]:
df['income'].isnull().sum()

0

Más adelante veremos si es necesario hacer limpieza de alguna columna más

In [21]:
# Guardamos nuestro set de datos tras la exploración inicial
df.to_csv('../Datos/EDA_vinos.csv')